{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "within genes and overlapping\n",
    "overlapping genes:\n",
    "in between genes\n",
    "found none genes\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "within_df = pd.read_csv('/home/natalia/dev/BRIDGe2023-main/dev/chromosome_files_mb_sc/chr1.csv')\n",
    "forward_exons_df = pd.read_csv('/home/natalia/dev/BRIDGe2023-main/dev/m_forward_sort_ex.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_columns = ['chromosome', 'read interval', 'reads', 'length of read','gene id','ambiguous', 'intronic', 'unspliced', 'spliced', \"3'UTR\",\"5'UTR\",'status' ]\n",
    "result_data = []\n",
    "\n",
    "def categorize_intervals(chrom, read_start, read_end, gene_id,forward_exons_df):\n",
    "\n",
    "    amb_result = []\n",
    "    intr_result = []\n",
    "    uns_result = []\n",
    "    spl_result = []\n",
    "    utr_3_result = []\n",
    "    utr_5_result = []\n",
    "    status = []\n",
    "\n",
    "    if pd.notna(gene_id):\n",
    "        same_gene = forward_exons_df[(forward_exons_df['chromosome'] == chrom) &\n",
    "                                    (forward_exons_df['gene_id'] == gene_id)]\n",
    "\n",
    "        if not same_gene.empty:\n",
    "            same_exon = same_gene[same_gene['transcript'] == same_gene['next transcript']]\n",
    "            if not same_exon.empty:\n",
    "                same_exon_number = same_exon[(same_exon['exon number']) < (same_exon['last exon number'])]\n",
    "                if not same_exon_number.empty:\n",
    "                    between_exons = same_exon_number[(same_exon_number['end'] < read_start) & (same_exon_number['next exon start'] > read_start)]\n",
    "                    if not between_exons.empty:\n",
    "                        #for i in range(len(intronic)):\n",
    "                        intronic = between_exons[(between_exons['next exon start'] > read_end)]\n",
    "                        if not intronic.empty:\n",
    "                            intr_exon_1 = intronic.iloc[0]['transcript']\n",
    "                            intr_exon_2 = intronic.iloc[0]['next transcript']\n",
    "                            intr_exon_number1 = intronic.iloc[0]['exon number']\n",
    "                            intr_exon_number2 = intronic.iloc[0]['next exon number']\n",
    "                            intr_result.append([[intr_exon_1, intr_exon_2], [intr_exon_number1, intr_exon_number2]])\n",
    "                            status.append(\"I\")\n",
    "                        unspliced = between_exons[(between_exons['next exon start'] < read_end) & (between_exons['next exon end'] > read_end)]\n",
    "                        if not unspliced.empty:\n",
    "                            #for i in range(len(unspliced)):\n",
    "                            uns_exon_1 = unspliced.iloc[0]['transcript']\n",
    "                            uns_exon_2 = unspliced.iloc[0]['next transcript']\n",
    "                            uns_exon_number1 = unspliced.iloc[0]['exon number']\n",
    "                            uns_exon_number2 = unspliced.iloc[0]['next exon number']\n",
    "                            uns_result.append([[uns_exon_1, uns_exon_2], [uns_exon_number1, uns_exon_number2]])\n",
    "                            status.append(\"U\")\n",
    "                            if 'N' in cigar:\n",
    "                                status.append('S')\n",
    "                    spliced = same_exon_number[(same_exon['start'] <= read_start) & (same_exon_number['end'] >= read_start) & (same_exon_number['next exon start'] <= read_end) & (same_exon_number['next exon end'] >= read_end) |\n",
    "                                            (same_exon['start'] <= read_start) & (same_exon_number['end'] >= read_start) & (same_exon_number['next exon start'] <= read_end) & (same_exon_number['next exon end'] <= read_end)]\n",
    "                    if not spliced.empty:\n",
    "                        #for i in range(len(spliced)):\n",
    "                        spl_1 = spliced.iloc[0]['transcript']\n",
    "                        spl_2 = spliced.iloc[0]['next transcript']\n",
    "                        spl_number_1 = spliced.iloc[0]['exon number']\n",
    "                        spl_number_2 = spliced.iloc[0]['next exon number']\n",
    "                        spl_result.append([[spl_1, spl_2], [spl_number_1, spl_number_2]])\n",
    "                        status.append(\"S\")\n",
    "                                \n",
    "                    unspliced = same_exon_number[(same_exon_number['end'] >= read_start) & (same_exon_number['start'] <= read_start) & (same_exon_number['next exon start'] > read_end) & (same_exon_number['end'] < read_end)]\n",
    "                    if not unspliced.empty:\n",
    "                        #for i in range(len(unspliced)):\n",
    "                        uns_exon_1 = unspliced.iloc[0]['transcript']\n",
    "                        uns_exon_2 = unspliced.iloc[0]['next transcript']\n",
    "                        uns_exon_number1 = unspliced.iloc[0]['exon number']\n",
    "                        uns_exon_number2 = unspliced.iloc[0]['next exon number']\n",
    "                        uns_result.append([[uns_exon_1, uns_exon_2], [uns_exon_number1, uns_exon_number2]])\n",
    "                        status.append(\"U\")\n",
    "                        if 'N' in cigar:\n",
    "                            status.append('S')\n",
    "                    if 'N' in cigar:\n",
    "                        status.clear()\n",
    "                        status.append(\"S\")\n",
    "            ambiguous = same_gene[(same_gene['start'] <= read_start) & (same_gene['end'] >= read_end)]\n",
    "            if not ambiguous.empty:\n",
    "                #for i in range(len(ambiguous)):            \n",
    "                amb_exon_name = ambiguous.iloc[0]['transcript']\n",
    "                amb_exon_number = ambiguous.iloc[0]['exon number']\n",
    "                amb_result.append([amb_exon_name, amb_exon_number])\n",
    "                status.clear()\n",
    "                status.append('A')\n",
    "            if 'N' in cigar:\n",
    "                        status.clear()\n",
    "                        status.append(\"S\")\n",
    "    else:\n",
    "        same_chrom = forward_exons_df[(forward_exons_df['chromosome'] == chrom)]\n",
    "        if not same_chrom.empty:\n",
    "            utr_5 = same_chrom[(same_chrom['first exon start'] >= read_end)]\n",
    "            if not utr_5.empty:\n",
    "                utr_5_exon_name = utr_5.iloc[0]['transcript']\n",
    "                utr_5_exon_number = utr_5.iloc[0]['exon number']\n",
    "                min_dist_5 = min(abs(utr_5['first exon start'] - read_end).min(), abs(utr_5['last exon end'] - read_start).min())\n",
    "                dist_to_gene_start = utr_5.iloc[0]['distance to gene start']\n",
    "                if min_dist_5 > (1000+dist_to_gene_start):\n",
    "                    status.append(\"None\")\n",
    "                else:\n",
    "                    utr_5_result.append([utr_5_exon_name, utr_5_exon_number])\n",
    "                    utr_5_result.append(f\"MinDist5:{min_dist_5}\")\n",
    "                    status.append(\"5T\")\n",
    "            else:\n",
    "                min_dist_5 = float('inf')\n",
    "            utr_3 = same_chrom[(same_chrom['last exon end'] <= read_start)]\n",
    "            if not utr_3.empty:\n",
    "                utr_3_exon_name = utr_3.iloc[0]['transcript']\n",
    "                utr_3_exon_number = utr_3.iloc[0]['exon number']\n",
    "                min_dist_3 = min(abs(utr_3['first exon start'] - read_start).min(), abs(utr_3['last exon end'] - read_start).min())\n",
    "                utr_3_result.append([utr_3_exon_name, utr_3_exon_number])\n",
    "                utr_3_result.append(f\"MinDist3:{min_dist_3}\")\n",
    "                status.append(\"3T\")\n",
    "                if \"None\" in status:\n",
    "                    status.clear()\n",
    "                    status.append(\"3T\")\n",
    "                elif \"5T\" in status:\n",
    "                    utr_3_result.clear()\n",
    "                    status.clear()\n",
    "                    status.append(\"5T\")\n",
    "            if 'N' in cigar:\n",
    "                utr_3_result.clear()\n",
    "                utr_5_result.clear()\n",
    "                status.clear()\n",
    "                status.append('S')\n",
    "\n",
    "    return [chrom, (read_start, read_end), reads, length_of_read, gene_id, amb_result, intr_result, uns_result, spl_result,utr_3_result,utr_5_result, status]\n",
    "\n",
    "# Iterate through within_df and process each row\n",
    "for index, read_interval_row in within_df.iterrows():\n",
    "    chrom = read_interval_row['chromosome']\n",
    "    read_start = read_interval_row['read start']\n",
    "    read_end = read_interval_row['read end']\n",
    "    reads = read_interval_row['reads']\n",
    "    gene_id = read_interval_row['gene_id']\n",
    "    cigar = read_interval_row['cigar string']\n",
    "    length_of_read = read_interval_row['length of read']\n",
    "\n",
    "    result_data.append(categorize_intervals(chrom, read_start, read_end, gene_id,forward_exons_df))\n",
    "    \n",
    "# Create a DataFrame using the result_columns and result_data\n",
    "result_df = pd.DataFrame(result_data, columns=result_columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df.to_csv('/home/natalia/dev/BRIDGe2023-main/dev/results.csv', index=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
